src/framework/ui/gsapiprovider.cpp

```diff
@@ -0,0 +1,346 @@
+/*
+ * SPDX-License-Identifier: GPL-3.0-only
+ * MuseScore-CLA-applies
+ *
+ * MuseScore
+ * Music Composition & Notation
+ *
+ * Copyright (C) 2024 MuseScore BVBA and others
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+#include "gsapiprovider.h"
+
+#include <QFile>
+#include <QQuickWindow>
+#include <QSGRendererInterface>
+
+#include "global/internal/globalconfiguration.h"
+#include "global/logger.h"
+
+#include "log.h"
+
+using namespace muse::ui;
+using namespace muse::logger;
+
+#ifdef Q_OS_WIN
+static const std::vector<GSApiProvider::GraphicsApi > ALLOWED_APIS = {
+    GSApiProvider::Direct3D11,
+    GSApiProvider::OpenGL,
+    GSApiProvider::Software
+};
+#endif
+
+#ifdef Q_OS_MACOS
+static const std::vector<GSApiProvider::GraphicsApi > ALLOWED_APIS = {
+    GSApiProvider::Metal,
+    GSApiProvider::OpenGL,
+    GSApiProvider::Software
+};
+#endif
+
+#ifdef Q_OS_LINUX
+static const std::vector<GSApiProvider::GraphicsApi > ALLOWED_APIS = {
+    GSApiProvider::OpenGL,
+    GSApiProvider::Software
+};
+#endif
+
+static const std::string BAD_MESSAGE = "Failed to build graphics pipeline state";
+GSTestObj* GSApiProvider::gsTestObj = nullptr;
+
+namespace muse::ui {
+class GSLogDest : public LogDest
+{
+public:
+
+    GSLogDest()
+        : LogDest(LogLayout("")) {}
+
+    std::string name() const { return "GlProblemsDetector"; }
+    void write(const LogMsg& logMsg)
+    {
+        if (isProblemDetected) {
+            return;
+        }
+
+        if (logMsg.message.find(BAD_MESSAGE) != std::string::npos) {
+            isProblemDetected = true;
+            std::cout << "found message: " << BAD_MESSAGE << std::endl;
+        }
+    }
+
+    bool isProblemDetected = false;
+};
+}
+
+GSApiProvider::GSApiProvider(const Version& appVersion)
+    : m_appVersion(appVersion)
+{
+    Logger* logger = Logger::instance();
+
+    m_logDest = new GSLogDest();
+    logger->addDest(m_logDest);
+
+    QObject::connect(&m_timer, &QTimer::timeout, [this]() {
```

Comment:
Alternative idea, instead of relying on a timer (for which we will need to specify a random guessed interval):
- let the GLLogDest take a callback
- when GLLogDest detects BAD_MESSAGE, it will call the callback
- the callback registers the current backend as broken and restarts the app
- when the callback has not been called at the point that the user closes the app manually, register the current backend as okay during the deinit step (or maybe it is not even necessary to explicitly register a backend as okay)

Reply:
The timer is needed mainly for two things:
* Perform a callback in the main thread (if we call a restart from another thread, there may be problems, because different actions are performed to exit the application)
* Do not do complex things in the log callback - this blocks the log and complicates testing due to some implementation features 

Instead of a timer, we can use a Qt meta object and make an invocation... (or just use an Async call, which is the same thing), but for this, the event loop needs to work... but it may not be running yet. That's why the timer.

Reply:
Timers also require the runloop to be working though, don't they? But it's fine, perhaps we shouldn't make big changes anymore now.

Reply:
I thought so too...
But the QMetaObject::invokeMethod call didn't work for me, and the timer works...
I didn't investigate why...
Just thoughts: either the timer works later, when the runloop has already been launched (maybe even this event is processed from some queue), or the timer doesn't need a runloop. After all, the timer is a system event, i.e. Qt makes a request to the system to set a timer, and then the system sends an event, the timer triggered... Although I don't know for sure.